File: SICP/Lecture_6_full/Code/Java/Stream/src/fayelab/sicp/stream/java8stream/ReQueens.java
package fayelab.sicp.stream.java8stream;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;

public class ReQueens
{
    public static Optional<List<List<List<Integer>>>> solveAll(int n)
    {
        List<List<List<Integer>>> solutions = fillRows(n, n).collect(toList());
        return solutions.isEmpty() ? Optional.empty() : Optional.of(solutions);
    }
    
    private static Stream<List<List<Integer>>> fillRows(int size, int row)
    {
        if(row == 0)
        {
            return Stream.of(emptyBoard());
        }
                
        return fillRows(size, row - 1)
                .flatMap(restPoses -> enumInterval(1, size).map(col -> adjoinPos(row, col, restPoses)))
                .filter(poses -> isSafe(poses));
    }
    
    public static Optional<List<List<List<Integer>>>> solveAll2(int n)
    {
        List<List<List<Integer>>> solutions = fillRows2(n, n).collect(toList());
        return solutions.isEmpty() ? Optional.empty() : Optional.of(solutions);
    }
    
    private static Stream<List<List<Integer>>> fillRows2(int size, int row)
    {
        if(row == 0)
        {
            return Stream.of(emptyBoard());
        }
                
        return enumInterval(1, size).flatMap(col -> fillRows2(size, row - 1).map(restPoses -> adjoinPos(row, col, restPoses)))
                                    .filter(poses -> isSafe(poses));
    }
    
    private static Stream<Integer> enumInterval(int low, int high)
    {
        return IntStream.rangeClosed(low, high).boxed();
    }

    private static List<List<Integer>> emptyBoard()
    {
        return new ArrayList<>();
    }

    private static List<List<Integer>> adjoinPos(int curRow, int curCol, List<List<Integer>> curPoses)
    {
        List<List<Integer>> result = new ArrayList<>();
        result.addAll(curPoses);
        result.add(asList(curRow, curCol));
        return result;
    }
    
    private static boolean isSafe(List<List<Integer>> poses)
    {
        List<Integer> newPos = poses.get(poses.size() - 1);
        List<List<Integer>> restPoses = poses.subList(0, poses.size() - 1);
        return isNotSameCol(newPos, restPoses) && isNotDiagonal(newPos, restPoses);
    }

    private static boolean isNotSameCol(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> newPos.get(1) == restPos.get(1));
    }

    private static boolean isNotDiagonal(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> Math.abs(newPos.get(0) - restPos.get(0)) 
                                                == Math.abs(newPos.get(1) - restPos.get(1)));
    }
}

File: SICP/Lecture_6_full/Code/Python/bif/requeens.py
## Eight Queens Problem, Recursive, Built-in functions
def solveall(n):
    solutions = fillrows(n, n)
    return solutions if solutions != [] else None

def fillrows(size, row):
    if row == 0:
        return [emptyboard()]

    return filter(lambda poses: issafe(poses), 
                  flatten2([adjoinpos(row, col, restposes) for col in range(1, size + 1) for restposes in fillrows(size, row - 1)]))

    # map(lambda col: adjoinpos(row, col, restposes), range(1, size + 1))

def emptyboard():
    return [];

def adjoinpos(currow, curcol, curposes):
    curposes.append((currow, curcol))

def issafe(poses):
    newpos = poses[-1]
    restposes = poses[:-1]
    return is_not_samecol(newpos, restposes) and is_not_diagonal(newpos, restposes)

def is_not_samecol(newpos, restposes):
    return all(map(lambda restpos: newpos[1] != restpos[1], restposes))

def is_not_diagonal(newpos, restposes):
    return all(map(lambda restpos: abs(newpos[0] - restpos[0]) != abs(newpos[1] - restpos[1]), restposes))

def flatten2(lst2layers):
    return [y for x in lst2layers for y in x]

## Tests
def test_solveall():
    assert [[(1, 2), (2, 4), (3, 1), (4, 3)],
            [(1, 3), (2, 1), (3, 4), (4, 2)]] == solveall(4)
    # assert None == solveall(3)
    # assert 92 == len(solveall(8))

    print 'test_solveall ok'

def test_solveall2():
    assert [[(1, 3), (2, 1), (3, 4), (4, 2)],
            [(1, 2), (2, 4), (3, 1), (4, 3)]] == solveall2(4)
    assert None == solveall2(3)
    # assert 92 == len(solveall(8))

    print 'test_solveall2 ok'    

def test():
    test_solveall()
    test_solveall2()

    print 'test ok'

if __name__ == '__main__':
    test()